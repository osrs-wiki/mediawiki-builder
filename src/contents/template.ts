import MediaWikiContent from "../content";

export type TemplateParam = { key: string; value: string };

export class MediaWikiTemplate extends MediaWikiContent {
  name: string;
  params: TemplateParam[];

  constructor(name: string) {
    super();
    this.name = name;
    this.params = [];
  }

  add(key: string, value: string) {
    this.params.push({ key, value });
  }

  build() {
    const expanded = this.params.length > 3;
    const params = this.params.reduce(
      (allParams, param) =>
        `${allParams}${this.params.length > 3 ? "\n" : ""}|${
          param.key ? param.key + (expanded ? " = " : "=") : ""
        }${param.value}`,
      ""
    );
    return `{{${this.name}${params}${expanded ? "\n" : ""}}}\n`;
  }
}
