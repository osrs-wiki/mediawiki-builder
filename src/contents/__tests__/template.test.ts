import { MediaWikiTemplate } from "../template";

describe("MediaWikiTemplate", () => {
  test("it should render flattened with less than 4 params", () => {
    const template = new MediaWikiTemplate("test");
    template.add("one", "one");
    template.add("two", "two");
    template.add("three", "three");

    expect(template.build()).toMatchSnapshot();
  });

  test("it should render expanded with 4 or more params", () => {
    const template = new MediaWikiTemplate("test");
    template.add("one", "one");
    template.add("two", "two");
    template.add("three", "three");
    template.add("four", "four");

    expect(template.build()).toMatchSnapshot();
  });
});
