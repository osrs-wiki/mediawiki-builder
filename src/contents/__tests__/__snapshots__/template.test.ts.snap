// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaWikiTemplate it should render expanded with 4 or more params 1`] = `
"{{test
|one = one
|two = two
|three = three
|four = four
}}
"
`;

exports[`MediaWikiTemplate it should render flattened with less than 4 params 1`] = `
"{{test|one=one|two=two|three=three}}
"
`;
